test: Do not pass tests on unhandled exceptions

maflcko opened the pull request:
Currently the functional tests are problematic, because they pass, even if they encounter an unhanded exception.

Fix this by handling all exceptions: Catch `BaseException` as fallback and mark it as failure.

Can be tested via:

```diff
diff --git a/test/functional/wallet_disable.py b/test/functional/wallet_disable.py
index da6e5d408f..ecc41fb041 100755
--- a/test/functional/wallet_disable.py
+++ b/test/functional/wallet_disable.py
@@ -19,6 +19,7 @@ class DisableWalletTest (BitcoinTestFramework):
         self.wallet_names = []
 
     def run_test (self):
+        import sys;sys.exit("fatal error")
         # Make sure wallet is really disabled
         assert_raises_rpc_error(-32601, 'Method not found', self.nodes[0].getwalletinfo)
         x = self.nodes[0].validateaddress('3J98t1WpEZ73CNmQviecrnyiWrnqRhWNLy')
```

Previously, the test would pass. With this patch, it would fail.

Commit:
test: Do not pass tests on unhandled exceptions

This adds a missing catch for BaseException (e.g. SystemExit), which
would otherwise be silently ignored.

Also, remove the redundant other catches, which are just calling
log.exception with a redundant log message.

stickies-v commented on test/functional/test_framework/test_framework.py:
```diff
@@ -215,6 +215,9 @@ def main(self):
             self.success = TestStatus.FAILED
+        except BaseException as e:
+            self.log.warning(f"Unhandled base exception {repr(e)}")
+            self.success = TestStatus.FAILED
```
Printing the exception `e` seems more useful than "Unexpected exception", and imo completely obviates the need for having the `CalledProcessError` and `KeyboardInterrupt` exceptions?

stickies-v reviewed (approved):
ACK fa30b34026f76a5b8af997152fced2d281782e0d

Commit:
test: Log KeyboardInterrupt as exception

log.exception is more verbose and useful to debug timeouts.

Also, log stderr for CalledProcessError to make debugging easier.

stickies-v reviewed (approved):
re-ACK faa3e684118bffa7a98cf76eeeb59243219df900

OrangeDoro reviewed (commented):
Hi! I'm a grad student working on a research project about using large language models to automate code review. Based on your commit faa3e684118bffa7a98cf76eeeb59243219df900 and the changes in test/functional/test_framework/test_framework.py, my tool generated this comment:
1. Ensure that `e.stdout` and `e.stderr` are checked for existence before logging to prevent potential `AttributeError`. Consider using `getattr(e, 'stdout', 'N/A')` and `getattr(e, 'stderr', 'N/A')`.
2. Change `except BaseException` to `except Exception` to avoid catching system-exiting exceptions unintentionally.
3. Ensure `self.success` is set to `TestStatus.FAILED` in the `BaseException` block to accurately reflect the test's outcome.
4. Ensure that the `shutdown` method in the `finally` block is well-defined and handles potential exceptions to enhance robustness.
5. Reintroduce the `KeyboardInterrupt` exception handling to ensure user interruptions are logged and handled properly.
6. Consider sanitizing the output before logging or avoid logging the entire output to prevent sensitive information exposure.
7. Ensure that the exit code handling is appropriate for the application context. Document the intended behavior if `sys.exit()` is called.
8. The assertion `assert hasattr(self, "num_nodes")` could be replaced with an explicit check and raise a `ValueError` if `self.num_nodes` is not set.
9. Consider logging the command that was executed in the `subprocess.CalledProcessError` block for better debugging context.
10. Add a test case that simulates a `subprocess.CalledProcessError` and verifies that the log message contains both `stdout` and `stderr`.
11. Capture the log output during tests and assert that it matches the expected format, including the presence of `stdout` and `stderr`.
12. The `self.success` assignments in the exception handling blocks could use comments to clarify what each status means in the context of the test execution.

As part of my research, I'm trying to understand how useful these comments are in real-world development. If you have a moment, I'd be super grateful if you could quickly reply to these two yes/no questions:


1. Does this comment provide suggestions from a dimension you hadn’t considered?

2. Do you find this comment helpful?

Thanks a lot for your time and feedback! And sorry again if this message is a bother.

pablomartin4btc reviewed (approved):
tACK faa3e684118bffa7a98cf76eeeb59243219df900

Managed to reproduce the issue with the patch provided in the PR description, this branch fixes it. Nice detail handling `KeyboardInterrupt` for debugging purpose. Code-wise refactoring: better reading removing redundant catches.

fanquake commented:
Backported to `29.x` in #33046.
